samcli/lib/deploy/deployer.py

```diff
@@ -285,39 +293,52 @@ def describe_stack_events(self, stack_name, time_stamp_marker, **kwargs):
-                            event["ResourceType"],
-                            event["LogicalResourceId"],
-                            event.get("ResourceStatusReason", "-"),
-                        ],
-                        width=kwargs["width"],
-                        margin=kwargs["margin"],
-                        format_string=DESCRIBE_STACK_EVENTS_FORMAT_STRING,
-                        format_args=kwargs["format_args"],
-                        columns_dict=DESCRIBE_STACK_EVENTS_DEFAULT_ARGS.copy(),
-                    )
-
-            if self._check_stack_complete(stack_status):
-                stack_change_in_progress = False
+                # Only sleep if there have been no retry_attempts
+                time.sleep(self.client_sleep if retry_attempts == 0 else 0)
+                describe_stacks_resp = self._client.describe_stacks(StackName=stack_name)
+                paginator = self._client.get_paginator("describe_stack_events")
+                response_iterator = paginator.paginate(StackName=stack_name)
+                stack_status = describe_stacks_resp["Stacks"][0]["StackStatus"]
+                for event_items in response_iterator:
+                    for event in event_items["StackEvents"]:
+                        if event["EventId"] not in events and utc_to_timestamp(event["Timestamp"]) > time_stamp_marker:
+                            events.add(event["EventId"])
+
+                            pprint_columns(
+                                columns=[
+                                    event["ResourceStatus"],
+                                    event["ResourceType"],
+                                    event["LogicalResourceId"],
+                                    event.get("ResourceStatusReason", "-"),
+                                ],
+                                width=kwargs["width"],
+                                margin=kwargs["margin"],
+                                format_string=DESCRIBE_STACK_EVENTS_FORMAT_STRING,
+                                format_args=kwargs["format_args"],
+                                columns_dict=DESCRIBE_STACK_EVENTS_DEFAULT_ARGS.copy(),
+                            )
+
+                if self._check_stack_complete(stack_status):
+                    stack_change_in_progress = False
+                    break
+            except botocore.exceptions.ClientError:
+                retry_attempts = retry_attempts + 1
+                if retry_attempts > self.max_attempts:
+                    raise
+                # Sleep in exponential backoff mode
+                time.sleep(math.pow(self.backoff, retry_attempts))
 
     def _check_stack_complete(self, status):
         return "COMPLETE" in status and "CLEANUP" not in status
 
     def wait_for_execute(self, stack_name, changeset_type):
 
-        click.secho("\nWaiting for stack create/update to complete\n")
+        sys.stdout.write("\nWaiting for stack create/update to complete\n")
```

Comment:
Are you sure these are written to stdout in AWS CLI?

Reply:
https://github.com/aws/aws-cli/blob/develop/awscli/customizations/cloudformation/deployer.py#L154

I can add flushes. 👍

Reply:
+1 add flushes